src/mds/MetricAggregator.cc

```diff
@@ -0,0 +1,340 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "MetricAggregator.h"
+#include "mgr/MgrClient.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_mds
+#undef dout_prefix
+#define dout_prefix *_dout << "mds.metric.aggregator" << " " << __func__
+
+MetricAggregator::MetricAggregator(CephContext *cct, MDSRank *mds, MgrClient *mgrc)
+  : Dispatcher(cct),
+    mgrc(mgrc),
+    mds_pinger(mds) {
+}
+
+void MetricAggregator::ping_all_active_ranks() {
+  dout(10) << ": pinging " << active.size() << " active mds(s)" << dendl;
+
+  for (auto &rank : active) {
+    dout(20) << ": pinging rank=" << rank << dendl;
+    mds_pinger.send_ping(rank);
```

Comment:
This races with updates to the MDSMap:

````
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr:terminate called after throwing an instance of 'std::out_of_range'
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr:  what():  map::at
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr:*** Caught signal (Aborted) **
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr: in thread 7f59411a8700 thread_name:ms_dispatch
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr: ceph version 15.0.0-9652-g0636e28 (0636e283d185bb5d812aee2ce73a767f34213264) octopus (dev)
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr: 1: (()+0x12dd0) [0x7f594f933dd0]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 2: (gsignal()+0x10f) [0x7f594e3f799f]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 3: (abort()+0x127) [0x7f594e3e1cf5]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 4: (()+0x8ff8b) [0x7f594edacf8b]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 5: (()+0x963ec) [0x7f594edb33ec]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 6: (()+0x96447) [0x7f594edb3447]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 7: (()+0x966a8) [0x7f594edb36a8]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 8: (()+0x91f2f) [0x7f594edaef2f]
2020-01-23T09:06:40.112 INFO:tasks.ceph.mds.a.smithi120.stderr: 9: (MDSRank::send_message_mds(boost::intrusive_ptr<Message> const&, int)+0x74e) [0x564c6327b8de]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 10: (MDSPinger::send_ping(int)+0x46d) [0x564c635cf95d]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 11: (MetricAggregator::ping_all_active_ranks()+0x195) [0x564c63560a25]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 12: (()+0x447c54) [0x564c63560c54]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 13: (()+0xc2a33) [0x7f594eddfa33]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 14: (()+0x82de) [0x7f594f9292de]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 15: (clone()+0x43) [0x7f594e4bc4b3]
````

From: /ceph/teuthology-archive/pdonnell-2020-01-23_00:39:49-kcephfs-wip-pdonnell-testing-20200122.210533-distro-basic-smithi/4695282/teuthology.log

I realize you considered this in the design but I'm afraid that this separate lock won't work. The call to `mds->send_message_mds` relies on `mds.mdsmap` which may be updated asynchronously.

This pinger thread needs to use the `mds_lock`. Or, alternatively, you can make a helper `MDSRank::send_message_mds` which is static and takes an MDSMap. Then store a copy of the MDSMap in `MetricAggregator` for passing to that helper. (I think that should work?)

Reply:
ack -- calls to `send_message_mds` can be wrapped around `mds_lock` -- in `MDSPinger::send-ping` and `MetricsHandler::update_rank0`. I think that should be ok -- the bulk of metrics processing would be outside `mds_lock`.

Reply:
> ack -- calls to `send_message_mds` can be wrapped around `mds_lock` -- in `MDSPinger::send-ping` and `MetricsHandler::update_rank0`. I think that should be ok -- the bulk of metrics processing would be outside `mds_lock`.

This will deadlock -- `MetricAggregator::notify_mdsmap` call lock order would be mds_lock -> locak_lock, but the pinger thread will ping active MDSs with local_lock -> MDSPinger::lock -> mds_lock (to keep minimum critical section under mds_lock). We could probably drop local_lock once a copy of the active set is made, but I think we are better off with maintaining a copy of mds map.

Reply:
> This pinger thread needs to use the `mds_lock`. Or, alternatively, you can make a helper `MDSRank::send_message_mds` which is static and takes an MDSMap. Then store a copy of the MDSMap in `MetricAggregator` for passing to that helper. (I think that should work?)

There *might* be other issues with the static `send_message_mds` approach -- currently `send_message_mds` additionally sends an updated MDSMap to its peer based on the cached epoch of the peer MDS (vs epoch from its mds map). How would the custom `send_message_mds` deal with this?

Right now, sequence numbers are sent from rank 0 to other active ranks and metrics are sent from all active ranks to rank 0 (carrying the sequence number). Ignoring to send updated mds map *might* just cause other problems.

Reply:
@batrick -- I think dropping local_lock before acquiring mds_lock would work. I'll give it a try tomorrow. Thoughts?

Reply:
Should work I think.

Reply:
> @batrick -- I think dropping local_lock before acquiring mds_lock would work. I'll give it a try tomorrow. Thoughts?

ERr, if you're going to pick up the mds_lock then what's the point of storing the set of active ranks?

Reply:
true. that will be a quirk in the design which has to be lived with until ->send_message_mds is decoupled from MDSRank (w/o the need to acquire mds_lock).